--[[
To explain, Forbidden has always strived to be FREE.
    And it always will be!

Each precoded solution must contain:
    - A "Lite" edition / Unpaid edition (can be limited features / not for publishing)
    - Documentation
    - A way for developers to gain access to the solution for free. (see below "# I cannot pay for the full version of Forbidden's Preprogrammed Solutions.")

Please read the whole list below, it should share with you the core concept behind each idea.  

The precoded solutions is my (@CritDEV) way of doing the following:
    - Supporting developers who have made their own NPC solution and want to share it.
    - Allowing developers to support the development of Forbidden.
    - Allowing developers to have access to basic systems to build their games for free.
    - Preventing developers from getting stuck on common pathfinding & NPC topics.
    - Providing more advanced solutions to those who need it.


# I cannot pay for the full version of Forbidden's Preprogrammed Solutions.
To get a precoded solution for free, provide a meaningful contribution to the Lite editions.
    Any feature / documentation / bug fix / resource contributed will not be put behind a pay-wall.
        > Please ask in the discord if your contribution / idea is valid before setting out to work.
        > I / A Creator of a Precoded is under no obligation to accept your solution if you do contact us initially. 
]]--