
--[[

# Overview

	Simple pathfinding module with tracking and precoded solutions.
		
		The module can track, yield the thread, and has various options.


# Troubleshooting

	Most of the time any issues that arise will be with Forbidden. But there are cases where common issues and misunderstandings 
	arise
	
		1. Make sure the target doesn't disappear (this actually happens alot, bc unanchored or other reasons)
		2. Make sure the terrain is actually passable.
		3. Terrain is a big issue with Roblox's pathfinding system so keep that in mind.
		4. Any lag that appears when the NPC goes close to a player is caused by the parts not doing "part:SetNetworkOwner(nil)" 
		The antilag included is not the best fix to this issue, just my attempt to mitigate this issue as much as possible for
		new developers.
		
		5. Forbidden DOES NOT support trusses.
		
		
		Any questions or issues please DM crit#0271 on discord.
	
# Functions

documentation soon :).
this is an old file from Forbidden V1
Moving all docs to github.

@rman501 (RBLX)  @CritDEV (YT)
]]--